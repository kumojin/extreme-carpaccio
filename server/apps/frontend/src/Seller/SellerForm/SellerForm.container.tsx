import type { AddSellerType } from '../Seller.hook';
import { SellerForm } from './SellerForm';
import { useForm } from './SellerFrom.hook';

type SellerFormContainerProps = {
  addSeller: AddSellerType;
};

const SellerFormContainer = ({ addSeller }: SellerFormContainerProps) => {
  const { nameRef, passwordRef, urlRef, handleSubmit, errorForm } =
    useForm(addSeller);

  return (
    <SellerForm
      nameRef={nameRef}
      passwordRef={passwordRef}
      urlRef={urlRef}
      handleSubmit={handleSubmit}
      errorForm={errorForm}
    />
  );
};

export default SellerFormContainer;
