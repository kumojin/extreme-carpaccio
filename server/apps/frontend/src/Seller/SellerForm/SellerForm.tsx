import { FormEvent, RefObject } from 'react';

type SellerFormProps = {
  handleSubmit: (event: FormEvent) => void;
  nameRef: RefObject<HTMLInputElement>;
  passwordRef: RefObject<HTMLInputElement>;
  urlRef: RefObject<HTMLInputElement>;
  isError: boolean;
};

export const SellerForm = ({
  handleSubmit,
  nameRef,
  passwordRef,
  urlRef,
  isError,
}: SellerFormProps): JSX.Element => (
  <div className="App">
    <div className="container-fluid text-sm-center p-5 mb-5 bg-light">
      <h2 className="p-3">Hello, Seller!</h2>
      <form onSubmit={handleSubmit}>
        <div className="row d-flex align-items-end justify-content-center">
          <div className="col-auto text-sm-start">
            <label htmlFor="name" className="sr-only p-1">
              Name
            </label>
            <input
              type="text"
              placeholder="your name"
              className="form-control"
              ref={nameRef}
              data-toggle="tooltip"
              data-placement="bottom"
              title="Your username"
            />
          </div>
          <div className="col-auto text-sm-start">
            <label htmlFor="password" className="sr-only p-1">
              Password
            </label>
            <input
              type="password"
              placeholder="your password"
              className="form-control"
              ref={passwordRef}
              data-toggle="tooltip"
              data-placement="bottom"
              title="Password is used if you want to register yourself on a different url. You will need to provide the same username with the same password. Beware that there is nothing that can be done to retrieve it..."
            />
          </div>
          <div className="col-auto text-sm-start">
            <label htmlFor="url" className="sr-only p-1">
              URL
            </label>
            <input
              type="text"
              placeholder="http://192.168.1.1:3000"
              className="form-control"
              ref={urlRef}
              data-toggle="tooltip"
              data-placement="bottom"
              title="Base url of your own client"
            />
          </div>
          <div className="col-auto align-self-end">
            <button type="submit" className="btn btn-success">
              Register
            </button>
          </div>
        </div>
      </form>
    </div>
    {isError ? (
      <div className="alert alert-danger" role="alert">
        Warning: There is an error.
      </div>
    ) : (
      <></>
    )}
  </div>
);
