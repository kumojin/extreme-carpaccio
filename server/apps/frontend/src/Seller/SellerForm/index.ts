export { default as SellerForm } from './SellerForm.container';
