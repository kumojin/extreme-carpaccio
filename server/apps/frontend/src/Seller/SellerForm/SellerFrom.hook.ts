import React, { FormEvent, RefObject } from 'react';
import { useMutation } from 'react-query';
import { AddSellerType, SellerForm } from '../Seller.hook';

export type ErrorFormType = {
  hasError: boolean;
  message: string;
};
export const useForm = (
  addSeller: AddSellerType
): {
  nameRef: RefObject<HTMLInputElement>;
  passwordRef: RefObject<HTMLInputElement>;
  urlRef: RefObject<HTMLInputElement>;
  handleSubmit: (event: FormEvent) => void;
  errorForm: ErrorFormType;
} => {
  const nameRef: RefObject<HTMLInputElement> = React.useRef(null);
  const passwordRef: RefObject<HTMLInputElement> = React.useRef(null);
  const urlRef: RefObject<HTMLInputElement> = React.useRef(null);
  const initErrorFormState = {
    hasError: false,
    message: '',
  };
  const [errorForm, setErrorForm] =
    React.useState<ErrorFormType>(initErrorFormState);

  const addSellerMutation = useMutation(async (newSeller: SellerForm) =>
    fetch('/seller', {
      method: 'POST',
      body: new URLSearchParams(newSeller),
    })
  );

  const handleSubmit = (event: FormEvent): void => {
    const name = nameRef.current?.value;
    const password = passwordRef.current?.value;
    const url = urlRef.current?.value;
    setErrorForm(initErrorFormState);

    event.preventDefault();
    if (!name) {
      setErrorForm({
        hasError: true,
        message: 'name is required',
      });
      return;
    }
    if (!password) {
      setErrorForm({
        hasError: true,
        message: 'password is required',
      });
      return;
    }
    if (!url) {
      setErrorForm({
        hasError: true,
        message: 'url is required',
      });
      return;
    }

    addSellerMutation.mutate(
      { name, password, url },
      {
        onSuccess: async (response) => {
          if (!response.ok) {
            setErrorForm({
              hasError: true,
              message: 'An error occurred during registration',
            });
          } else {
            addSeller({ name, cash: 0, online: true });
          }
        },
        onError: async (err) => {
          setErrorForm({
            hasError: true,
            message: `An error occured during registration: ${err}`,
          });
        },
      }
    );

    nameRef.current.value = '';
    urlRef.current.value = '';
    passwordRef.current.value = '';
  };

  return { nameRef, passwordRef, urlRef, handleSubmit, errorForm };
};
