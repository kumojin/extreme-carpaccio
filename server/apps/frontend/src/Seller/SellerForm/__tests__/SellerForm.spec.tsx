import { useRef } from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import '@testing-library/jest-dom';
import { SellerForm } from '../SellerForm';

jest.mock('react', () => {
  const originReact = jest.requireActual('react');
  const mUseRef = jest.fn();
  return {
    ...originReact,
    useRef: mUseRef,
  };
});

describe('SellerForm', () => {
  const handleSubmit = jest.fn();
  const nameRef = useRef(null);
  const passwordRef = useRef(null);
  const urlRef = useRef(null);

  test('should render shellerForm without error', () => {
    // ARRANGE
    const { baseElement } = render(
      <SellerForm
        handleSubmit={handleSubmit}
        nameRef={nameRef}
        passwordRef={passwordRef}
        urlRef={urlRef}
        isError={false}
      />
    );
    expect(baseElement).toMatchSnapshot();
  });

  test('should render shellerForm with error', () => {
    // ARRANGE
    const { baseElement } = render(
      <SellerForm
        handleSubmit={handleSubmit}
        nameRef={nameRef}
        passwordRef={passwordRef}
        urlRef={urlRef}
        isError={true}
      />
    );
    expect(baseElement).toMatchSnapshot();
    expect(screen.getByRole('alert')).toHaveTextContent(
      'Warning: There is an error.'
    );
  });

  test('should handleSubmit is called when I click on the button', async () => {
    render(
      <SellerForm
        handleSubmit={handleSubmit}
        nameRef={nameRef}
        passwordRef={passwordRef}
        urlRef={urlRef}
        isError={true}
      />
    );

    await userEvent.click(screen.getByText('Register'));
    expect(handleSubmit).toBeCalled();
  });
});
