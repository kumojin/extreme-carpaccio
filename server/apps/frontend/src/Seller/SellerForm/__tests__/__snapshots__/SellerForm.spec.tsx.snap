// Vitest Snapshot v1

exports[`SellerForm > should render shellerForm with error 1`] = `
<body>
  <div>
    <div
      class="App"
    >
      <div
        class="container-fluid text-sm-center p-5 mb-5 bg-light"
      >
        <h2
          class="p-3"
        >
          Hello, Seller!
        </h2>
        <form>
          <div
            class="row d-flex align-items-end justify-content-center"
          >
            <div
              class="col-auto text-sm-start"
            >
              <label
                class="sr-only p-1"
                for="name"
              >
                Name
              </label>
              <input
                aria-label="your username"
                class="form-control"
                data-placement="bottom"
                data-toggle="tooltip"
                id="name"
                placeholder="your name"
                required=""
                title="Your username"
                type="text"
              />
            </div>
            <div
              class="col-auto text-sm-start"
            >
              <label
                class="sr-only p-1"
                for="password"
              >
                Password
              </label>
              <input
                aria-label="your password"
                class="form-control"
                data-placement="bottom"
                data-toggle="tooltip"
                id="password"
                placeholder="your password"
                required=""
                title="Password is used if you want to register yourself on a different url. You will need to provide the same username with the same password. Beware that there is nothing that can be done to retrieve it..."
                type="password"
              />
            </div>
            <div
              class="col-auto text-sm-start"
            >
              <label
                class="sr-only p-1"
                for="url"
              >
                URL
              </label>
              <input
                aria-label="your url"
                class="form-control"
                data-placement="bottom"
                data-toggle="tooltip"
                id="url"
                placeholder="http://192.168.1.1:3000"
                required=""
                title="Base url of your own client"
                type="text"
              />
            </div>
            <div
              class="col-auto align-self-end"
            >
              <button
                aria-label="register-button"
                class="btn btn-success"
                type="submit"
              >
                Register
              </button>
            </div>
          </div>
        </form>
      </div>
      <div
        class="alert alert-danger"
        role="alert"
      >
        Warning: There is an error.
      </div>
    </div>
  </div>
</body>
`;

exports[`SellerForm > should render shellerForm without error 1`] = `
<body>
  <div>
    <div
      class="App"
    >
      <div
        class="container-fluid text-sm-center p-5 mb-5 bg-light"
      >
        <h2
          class="p-3"
        >
          Hello, Seller!
        </h2>
        <form>
          <div
            class="row d-flex align-items-end justify-content-center"
          >
            <div
              class="col-auto text-sm-start"
            >
              <label
                class="sr-only p-1"
                for="name"
              >
                Name
              </label>
              <input
                aria-label="your username"
                class="form-control"
                data-placement="bottom"
                data-toggle="tooltip"
                id="name"
                placeholder="your name"
                required=""
                title="Your username"
                type="text"
              />
            </div>
            <div
              class="col-auto text-sm-start"
            >
              <label
                class="sr-only p-1"
                for="password"
              >
                Password
              </label>
              <input
                aria-label="your password"
                class="form-control"
                data-placement="bottom"
                data-toggle="tooltip"
                id="password"
                placeholder="your password"
                required=""
                title="Password is used if you want to register yourself on a different url. You will need to provide the same username with the same password. Beware that there is nothing that can be done to retrieve it..."
                type="password"
              />
            </div>
            <div
              class="col-auto text-sm-start"
            >
              <label
                class="sr-only p-1"
                for="url"
              >
                URL
              </label>
              <input
                aria-label="your url"
                class="form-control"
                data-placement="bottom"
                data-toggle="tooltip"
                id="url"
                placeholder="http://192.168.1.1:3000"
                required=""
                title="Base url of your own client"
                type="text"
              />
            </div>
            <div
              class="col-auto align-self-end"
            >
              <button
                aria-label="register-button"
                class="btn btn-success"
                type="submit"
              >
                Register
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</body>
`;
